torch/testing/_internal/common_distributed.py

```diff
@@ -167,31 +168,28 @@ class MultiProcessTestCase(TestCase):
     def world_size(self):
         return 4
 
-    @staticmethod
-    def join_or_run(fn):
+    def join_or_run(self, fn):
         @wraps(fn)
         def wrapper(self):
             if self.rank == self.MAIN_PROCESS_RANK:
                 self._join_processes(fn)
             else:
                 try:
-                    fn(self)
+                    fn()
                 except Exception as e:
                     logging.error('Caught exception: \n{}exiting process with exit code: {}'
                                   .format(traceback.format_exc(), MultiProcessTestCase.TEST_ERROR_EXIT_CODE))
                     sys.exit(MultiProcessTestCase.TEST_ERROR_EXIT_CODE)
-        return wrapper
+        return types.MethodType(wrapper, self)
 
     # The main process spawns N subprocesses that run the test.
     # This function patches overwrites every test function to either
     # assume the role of the main process and join its subprocesses,
     # or run the underlying test function.
```

Comment:
Are you sure this comment is up to date? I think it would be pretty helpful to have a description of what exactly step-by-step happens dynamic execution wise when you start the parent process, and when you hit the child process. One of the reasons I wanted @mrshenli to have a look is I don't actually know how the previous code used to work, so unfortunately, you're going to have to explain it to me :)

Reply:
Good point, let me update the comment to say, that it wraps only current test function.

Reply:
Sorry about the delay. This comments still look correct to me (although we should remove either "patches" or "overwrites"). 

Here is how it works. Let's take `test_rpc_spawn.py` as an example. Its `setUp` function calls into `_spawn_processes()` implemented in this file.  Each distributed test employs one main process and `world_size` sub-processes, where `world_size=4` by default. They can be distinguished by their ranks, where main process has a non initialized rank (-1) and sub process ranks ranging from 0 to 3. Main process and sub-process behave differently. Main process launches sub-processes and then waits for them to join and subprocesses runs the actual test function. And hence we have this `setUpClass` here.

Reply:
Yeah, I would love to see this comment in the code. I think it's kind of amazing how the subprocesses are able to figure out which test functions they should run haha.

Reply:
Sure, will add more comments when this is landed.

Reply:
Comments added in #37179